### PR TITLE
:warning: clusterctl v2 add the low-level library for managing provider's repository using Github

### PR DESCRIPTION
**Pull request by jiatongw**
**What this PR does / why we need it**:

This PR implements first backend implementation for repository using Github

- Support for reading from Github releases
- Not support for reading from repository source tree

**Which issue(s) this PR fixes** :
Rif #1729

/assign @fabriziopandini 
/cc @detiber @ncdc

### CODE REVIEW

**Comment by jiatongw** [comment]
@fabriziopandini Addressed comments

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 1]
```diff
@@ -0,0 +1,287 @@
+/*
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/repository/repository_github.go]: Please add comments to each method explaining what it does and why

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+	if configVariablesClient == nil {
+		return nil, errors.New("invalid arguments: configVariablesClient can't be nil")
+	}
+
+	rURL, err := url.Parse(providerConfig.URL())
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid url")
+	}
+
+	// if the url is a github repository
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/repository/repository_github.go]: Nit: Use complete sentences for inline comments https://golang.org/doc/effective_go.html#commentary

Include also the reasoning of a code line

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+const (
+	httpsScheme              = "https"
+	githubDomain             = "github.com"
+	gitHubTokeVariable       = "github-token"
+	gitHubReleaseRepository  = "releases"
+	gitHubLatestReleaseLabel = "latest"
+)
+
+// gitHubRepository provide support for providers hosted on GitHub.
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/repository/repository_github.go]:
```suggestion
// gitHubRepository provides support for providers hosted on GitHub.
```

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+//
+// We support for GitHub repository that reading from GitHub releases.
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/repository/repository_github.go]:
```suggestion
// We support GitHub repositories that use the release feature to publish artifacts and versions.
```

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+// Repository has full support for version, including also "latest" meta version.
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/repository/repository_github.go]:
```suggestion
// Repositories must use versioned releases, including the "latest" meta version (https://help.github.com/en/github/administering-a-repository/linking-to-releases#linking-to-the-latest-release).
```

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+type gitHubRepository struct {
+	providerConfig           config.Provider
+	configVariablesClient    config.VariablesClient
+	authenticatingHTTPClient *http.Client
+	owner                    string
+	repository               string //this is the GitHub repository
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/repository/repository_github.go]:
```suggestion
	repository               string
```

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+	// extracts all the info from path
+	owner := t[0]
+	repository := t[1]
+	defaultVersion := t[3]
+	path := strings.Join(t[4:], "/")
+
+	// use path's directory as a rootPath
+	rootPath := filepath.Dir(path)
+	// use the file name (if any) as componentsPath
+	componentsPath := strings.TrimPrefix(strings.TrimPrefix(path, rootPath), "/")
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/repository/repository_github.go]: This is a little hard to follow, I'd suggest to make it a method and have some unit tests separately for it

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+		providerConfig:        providerConfig,
+		configVariablesClient: configVariablesClient,
+		owner:                 owner,
+		repository:            repository,
+		defaultVersion:        defaultVersion,
+		rootPath:              rootPath,
+		componentsPath:        componentsPath,
+	}
+
+	token, err := configVariablesClient.Get(gitHubTokeVariable)
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/repository/repository_github.go]:
```suggestion
	token, err := configVariablesClient.Get(githubTokenVariable)
```

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+	if err != nil {
+		klog.V(1).Infof("The %q configuration variable is missing. Failing back to unauthenticated requests that allows for up to 60 requests per hour.", gitHubTokeVariable)
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/repository/repository_github.go]:
```suggestion
		klog.V(1).Infof("The %q configuration variable is missing. Falling back to unauthenticated requests that allows for up to 60 requests per hour.", gitHubTokenVariable)
```

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+func (g *gitHubRepository) getClient() *github.Client {
+	if g.injectClient != nil {
+		return g.injectClient
+	}
+	return github.NewClient(g.authenticatingHTTPClient)
+}
+
+func (g *gitHubRepository) setAuthenticatingClient(token string) {
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/repository/repository_github.go]:
```suggestion
func (g *gitHubRepository) setClientToken(token string) {
```

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+	ctx := context.Background()
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/repository/repository_github.go]: Context should be passed in as variable

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token},
+	)
+	g.authenticatingHTTPClient = oauth2.NewClient(ctx, ts)
+}
+
+// getLatestRelease return the latest release for a github repository, according to
+// semantic version order of the release tag name.
+func (g *gitHubRepository) getLatestRelease() (string, error) {
+	ctx := context.Background()
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/repository/repository_github.go]: Pass in as input

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+		return content, nil
+	}
+
+	// handle the case when it is returned a redirect link for a release asset
+	resp, err := http.Get(red) //nolint
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to download file %q from %q release via redirect location %q", *release.TagName, fileName, red)
+	}
+	defer resp.Body.Close()
+	content, err := ioutil.ReadAll(resp.Body)
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/repository/repository_github.go]: Leave space above this line for readability

**Comment by fabriziopandini** [comment]
@jiatongw #1891 has merged, you can now rebase

File: cmd/clusterctl/pkg/client/repository/repository_github.go (hunk 7 above)
Comment (jiatongw) [reply cmd/clusterctl/pkg/client/repository/repository_github.go]: Added a new method `getComponentsPath`, tests will do in a follow PR

**Comment by jiatongw** [comment]
@vincepri comments addressed!

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+	// extracts all the info from path
+	owner := t[0]
+	repository := t[1]
+	defaultVersion := t[3]
+	path := strings.Join(t[4:], "/")
+
+	// use path's directory as a rootPath
+	rootPath := filepath.Dir(path)
+	// use the file name (if any) as componentsPath
+	componentsPath := getComponentsPath(path, rootPath)
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/repository/repository_github.go]: `getComponentsPath` was introduced to address a comment above and to avoid a piece of ugly code; there is also a dedicated unit test now. IMO we should not rollback to inline expressions.

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+	// Check if the url is a github repository
+	if !(rURL.Scheme == httpsScheme && rURL.Host == githubDomain) {
+		return nil, errors.New("invalid url: a GitHub repository url should start with https://github.com")
+	}
+
+	// Check if the repository Path is in the expected format.
+	t := strings.Split(strings.TrimPrefix(rURL.Path, "/"), "/")
+	if len(t) < 5 || t[2] != githubReleaseRepository {
+		return nil, errors.Errorf(
+			"invalid url: a GitHub repository url should be in the form https://github.com/{owner}/{Path}/%s/{latest|version-tag}/{componentsClient.yaml}",
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/repository/repository_github.go]: Ok to rename Path into Project or Repository (I'm +1 on Repository)
I'm -1 to the second change because {path/components.yaml} does not reflect the correct order

File: cmd/clusterctl/pkg/client/repository/repository_github.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (jiatongw) [reply cmd/clusterctl/pkg/client/repository/repository_github.go]: https://github.com/kubernetes-sigs/cluster-api/pull/1909#discussion_r358896073
here is the previous comment about this change

File: cmd/clusterctl/pkg/client/repository/repository_github.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (jiatongw) [reply cmd/clusterctl/pkg/client/repository/repository_github.go]: Addressed

**Comment by fabriziopandini** [comment]
@jiatongw FYI https://github.com/kubernetes-sigs/cluster-api/pull/1978 is going to change the Repository interface; hopefully, this should not be a big change for your PR (everything you need is already implemented in getLatestRelease). 
I'm also open to get this PR merged as it is and make the required changes into #1978

**Comment by jiatongw** [comment]
@fabriziopandini rebased

**Comment by vincepri** [comment]
Seems the build is failing here

File: cmd/clusterctl/pkg/client/repository/repository_github.go [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+func (g *gitHubRepository) getVersions() ([]string, error) {
+	client := g.getClient()
+
+	// get all the releases
+	// NB. currently Github API does not support result ordering, so it not possible to limit results
+	releases, _, err := client.Repositories.ListReleases(context.Background(), g.owner, g.repository, nil)
+	if err != nil {
+		return nil, g.handleGithubErr(err, "failed to get the list of releases")
+	}
+	var versions []string //nolint
```
Comment (jiatongw) [inline cmd/clusterctl/pkg/client/repository/repository_github.go]: nolint here because we don't know the length in advance.

**Comment by jiatongw** [comment]
@vincepri PR updated after Repository interface changes. (Added `GetVersions` method and a helper method)

File: cmd/clusterctl/pkg/client/repository/repository_github.go (hunk 16 above)
Comment (vincepri) [reply cmd/clusterctl/pkg/client/repository/repository_github.go]: You can try `versions := []string{}`

Comment (jiatongw) [reply cmd/clusterctl/pkg/client/repository/repository_github.go]: Thanks! Done

**Comment by fabriziopandini** [comment]
@jiatongw, please squash commits and then this is ready to go for me
/approve

I'm looking forward to testing this with init as soon as both PR merges 😉

**Comment by jiatongw** [comment]
@fabriziopandini Done!

**Comment by fabriziopandini** [comment]
@jiatongw thanks!
/lgtm
/approve
